File: planner_core/validation.py
import re


WORK_HOURS_PATTERN = re.compile(r"^([01]?\d|2[0-3])-([01]?\d|2[0-3])$")
TIME_PATTERN = re.compile(r"^(?:[01]\d|2[0-3]):[0-5]\d$")


def parse_work_hours(work_hours: str) -> tuple[int, int]:
    match = WORK_HOURS_PATTERN.fullmatch(work_hours)
    if not match:
        raise ValueError("work_hours must use the format H-H or HH-HH within 0-23")

    start_h = int(match.group(1))
    end_h = int(match.group(2))
    if start_h >= end_h:
        raise ValueError("work_hours must have a start hour earlier than the end hour")

    return start_h * 60, end_h * 60


def parse_clock_time(value: str, field_name: str = "time") -> int:
    if not TIME_PATTERN.fullmatch(value):
        raise ValueError(f"{field_name} must use the format HH:MM")

    hours, minutes = value.split(":")
    return int(hours) * 60 + int(minutes)


def validate_time_blocks(
    blocks: list[dict],
    work_hours: str,
    current_time: str | None = None,
) -> list[dict]:
    if not blocks:
        raise ValueError("Expected at least one time block")

    day_start, day_end = parse_work_hours(work_hours)
    current_minutes = parse_clock_time(current_time, "current_time") if current_time else None
    previous_end = None

    for index, block in enumerate(blocks):
        task_title = block["task_title"].strip()
        if not task_title:
            raise ValueError(f"blocks[{index}].task_title cannot be empty")

        start_minutes = parse_clock_time(block["start"], f"blocks[{index}].start")
        end_minutes = parse_clock_time(block["end"], f"blocks[{index}].end")

        if start_minutes >= end_minutes:
            raise ValueError(f"blocks[{index}] must end after it starts")
        if start_minutes < day_start:
            raise ValueError(f"blocks[{index}] starts before the work window")
        if end_minutes > day_end:
            raise ValueError(f"blocks[{index}] ends after the work window")
        if current_minutes is not None and start_minutes < current_minutes:
            raise ValueError(f"blocks[{index}] starts before current_time")
        if previous_end is not None and start_minutes < previous_end:
            raise ValueError(f"blocks[{index}] overlaps a previous block")

        previous_end = end_minutes

    return blocks

File: backend/src/validation.ts
import { z } from "zod";

import type { RemainingTask, Task, TimeBlock } from "./types";

const workHoursPattern = /^([01]?\d|2[0-3])-([01]?\d|2[0-3])$/;
const timePattern = /^(?:[01]\d|2[0-3]):[0-5]\d$/;

export function parseWorkHours(workHours: string): [number, number] {
  const match = workHours.match(workHoursPattern);
  if (!match) {
    throw new Error("work_hours must use the format H-H or HH-HH within 0-23");
  }

  const startHour = Number(match[1]);
  const endHour = Number(match[2]);
  if (startHour >= endHour) {
    throw new Error("work_hours must have a start hour earlier than the end hour");
  }

  return [startHour * 60, endHour * 60];
}

export function parseClockTime(value: string, fieldName = "time"): number {
  if (!timePattern.test(value)) {
    throw new Error(`${fieldName} must use the format HH:MM`);
  }

  const [hours, minutes] = value.split(":").map(Number);
  return hours * 60 + minutes;
}

export function validateTimeBlocks(
  blocks: TimeBlock[],
  workHours: string,
  currentTime?: string,
): TimeBlock[] {
  if (!blocks.length) {
    throw new Error("Expected at least one time block");
  }

  const [dayStart, dayEnd] = parseWorkHours(workHours);
  const currentMinutes = currentTime ? parseClockTime(currentTime, "current_time") : undefined;
  let previousEnd: number | undefined;

  for (const [index, block] of blocks.entries()) {
    if (!block.task_title.trim()) {
      throw new Error(`blocks[${index}].task_title cannot be empty`);
    }

    const startMinutes = parseClockTime(block.start, `blocks[${index}].start`);
    const endMinutes = parseClockTime(block.end, `blocks[${index}].end`);

    if (startMinutes >= endMinutes) {
      throw new Error(`blocks[${index}] must end after it starts`);
    }
    if (startMinutes < dayStart) {
      throw new Error(`blocks[${index}] starts before the work window`);
    }
    if (endMinutes > dayEnd) {
      throw new Error(`blocks[${index}] ends after the work window`);
    }
    if (currentMinutes !== undefined && startMinutes < currentMinutes) {
      throw new Error(`blocks[${index}] starts before current_time`);
    }
    if (previousEnd !== undefined && startMinutes < previousEnd) {
      throw new Error(`blocks[${index}] overlaps a previous block`);
    }

    previousEnd = endMinutes;
  }

  return blocks;
}

export const taskSchema: z.ZodType<Task> = z.object({
  id: z.string().optional(),
  title: z.string().min(1),
  estimate_minutes: z.number().int().min(15).max(240),
  energy: z.enum(["low", "med", "high"]),
  dependencies: z.array(z.string()),
  required_materials: z.array(z.string()),
  acceptance_criteria: z.string().min(1),
  risk_blockers: z.array(z.string()),
});

export const remainingTaskSchema: z.ZodType<RemainingTask> = z.object({
  task_title: z.string().min(1),
  mode: z.enum(["deep", "light", "admin"]),
});

export const timeBlockSchema: z.ZodType<TimeBlock> = z.object({
  task_title: z.string().min(1),
  start: z.string(),
  end: z.string(),
  mode: z.enum(["deep", "light", "admin", "break"]),
});

export const decomposeRequestSchema = z.object({
  goal: z.string().min(1),
  work_hours: z.string().default("9-18"),
  deadline: z.string().nullable().optional(),
});

export const scheduleRequestSchema = z.object({
  tasks: z.array(taskSchema).min(1),
  work_hours: z.string().default("9-18"),
  date: z.string().nullable().optional(),
});

export const rescheduleRequestSchema = z.object({
  remaining_tasks: z.array(remainingTaskSchema).min(1),
  current_time: z.string(),
  work_hours: z.string().default("9-18"),
  state_signal: z.record(z.any()).nullable().optional(),
});

export function normalizeWorkHours(workHours: string): string {
  parseWorkHours(workHours);
  return workHours;
}

export function normalizeCurrentTime(currentTime: string): string {
  parseClockTime(currentTime, "current_time");
  return currentTime;
}
